R/tmb-methods.R

```diff
@@ -40,43 +40,146 @@ fitted.mmrm_tmb <- function(object, ...) {
+    interval = c("none", "confidence", "prediction"), level = 0.95,
+    na.action = na.pass, n_sim = 1000L, ...) { # nolint
+  interval <- match.arg(interval)
+  # make sure new data has the same levels as original data
+  full_frame <- model.frame(
+    object,
+    data = newdata,
+    include = c("subject_var", "visit_var", "group_var", "response_var"),
+    na.action = "na.pass"
+  )
+  tmb_data <- h_mmrm_tmb_data(
+    object$formula_parts, full_frame,
+    weights = rep(1, nrow(full_frame)), reml = TRUE,
+    singular = "keep", drop_visit_levels = FALSE, allow_na_response = TRUE, drop_levels = FALSE
+  )
+  predictions <- h_get_prediction(tmb_data, object$theta_est, object$beta_est, object$beta_vcov)
+  res <- data.frame(fit = predictions[, 1])
+  se <- switch(interval,
+    "confidence" = sqrt(predictions[, 2]),
+    "prediction" = sqrt(h_get_prediction_variance(object, n_sim, tmb_data)),
+    "none" = NULL
+  )
+  if (se.fit && interval != "none") { # save se
+    res <- cbind(res, se = se)
+  }
+  if (interval != "none") { # compute confidence interval
+    alpha <- 1 - level
+    z <- qnorm(1 - alpha / 2) * se
+    res <- cbind(res,
+      lwr = res[, "fit"] - z,
+      upr = res[, "fit"] + z
+    )
+  }
+  if (ncol(res) == 1) { # return vector if only fit is computed
+    res <- res[, "fit"]
+  }
+  return(res)
+}
+
+#' Get Prediction
+#' @description Get prediction with given data, theta, beta, beta_vcov
+#'
+#' @param tmb_data (`mmrm_tmb_data`)\cr object.
+#' @param theta (`numeric`)\cr theta value.
+#' @param beta (`numeric`)\cr beta value.
+#' @param beta_vcov (`matrix`)\cr beta_vcov matrix.
+#'
+#' @keywords internal
+h_get_prediction <- function(tmb_data, theta, beta, beta_vcov) {
+  assert_class(tmb_data, "mmrm_tmb_data")
+  assert_numeric(theta)
+  n_beta <- ncol(tmb_data$x_matrix)
+  assert_numeric(beta, finite = TRUE, any.missing = FALSE, len = n_beta)
+  assert_matrix(beta_vcov, mode = "numeric", any.missing = FALSE, nrows = n_beta, ncols = n_beta)
+  .Call(`_mmrm_predict`, PACKAGE = "mmrm", tmb_data, theta, beta, beta_vcov)
+}
+
+#' Get Prediction Variance
+#' @description Get prediction variance with given fit, tmb_data with sampling method
+#'
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param n_sim (`integer`)\cr number of replication of sampling.
+#' @param tmb_data (`mmrm_tmb_data`)\cr object.
+#'
+#' @keywords internal
+h_get_prediction_variance <- function(object, n_sim, tmb_data) {
+  assert_class(object, "mmrm_tmb")
+  assert_class(tmb_data, "mmrm_tmb_data")
+  assert_int(n_sim)
+  theta_chol <- chol(object$theta_vcov)
+  n_theta <- length(object$theta_est)
+  res <- replicate(n_sim, {
+    z <- rnorm(n = n_theta)
+    theta_sample <- object$theta_est + theta_chol %*% z
+    cond_beta_results <- object$tmb_object$report(theta_sample)
+    beta_mean <- cond_beta_results$beta
+    beta_cov <- cond_beta_results$beta_vcov
+    h_get_prediction(tmb_data, theta_sample, beta_mean, beta_cov)
```

Comment:
did you check whether we might be able to get the Cholesky factor of beta_cov directly from TMB reported? so that we don't need a new Cholesky decomposition in each sampling iteration?

Reply:
there is no cholesky decomposition for each sampling iteration; however, there is some cholesky decomposition for each sigma (to construct the sigma from theta need cholesky, and that is what I mentioned previously (which could, also reduce the computation time for autodiff as not cholesky is needed anymore)